invoice/invoice.go

```
@@ -0,0 +1,966 @@
+	// can fit into fewer groups we add leading zero groups, if it is too
+	// big we fail early, as there is not possible to encode it.
+	if len(timestampBase32) > 7 {
+		return "", fmt.Errorf("timestamp too big: %d", invoice.Timestamp)
+	}
+
+	// Add zero bytes to the first 7-len(timestampBase32) groups, then add
+	// the non-zero groups.
+	zeroes := make([]byte, 7-len(timestampBase32), 7-len(timestampBase32))
+	bufferBase32.Write(zeroes)
+	bufferBase32.Write(timestampBase32)
+
+	// We now write the tagged fields to the buffer, which will fill the
+	// rest of the data part before the signature.
+	if err := writeTaggedFields(&bufferBase32, invoice); err != nil {
+		return "", err
+	}
+
+	// The human-readable part (hrp) is "ln" + net hrp + optional amount.
+	hrp := "ln" + invoice.Net.Bech32HRPSegwit
+	if invoice.MilliSat != nil {
+		// Encode the amount using the fewest possible characters.
+		am, err := encodeAmount(*invoice.MilliSat)
+		if err != nil {
+			return "", err
+		}
+		hrp += am
+	}
+
+	// The signature is over the single SHA-256 hash of the hrp + the
+	// tagged fields encoded in base256.
+	taggedFieldsBytes, err := bech32.ConvertBits(bufferBase32.Bytes(), 5, 8, true)
+	if err != nil {
+		return "", err
+	}
+
+	toSign := append([]byte(hrp), taggedFieldsBytes...)
+	hash := chainhash.HashB(toSign)
+
+	// We use compact signature format, and also encoded the recovery ID
+	// such that a reader of the invoice can recover our pubkey from the
+	// signature.
+	sign, err := signer.SignCompact(hash)
+	if err != nil {
+		return "", err
+	}
+
+	// From the header byte we can extract the recovery ID, and the last 64
+	// bytes encode the signature.
+	recoveryID := sign[0] - 27 - 4
+	var sigBytes [64]byte
+	copy(sigBytes[:], sign[1:])
+
+	// If the pubkey field was explicitly set, it must be set to the pubkey
+	// used to create the signature.
+	if invoice.PubKey != nil {
+		var signature *btcec.Signature
+		lnwire.DeserializeSigFromWire(&signature, sigBytes)
```

Ignoring an error here.